merkletree.go

```diff
@@ -0,0 +1,344 @@
+
+			currentNodeUL.index = currentNodeKey
+
+			direction := getNthBit(currentNodeKey, position)
+			if direction {
+				newInteriorNode.rightChild = currentNodeUL
+			} else {
+				newInteriorNode.leftChild = currentNodeUL
+			}
+			currentNodeUL.level++
+			currentNodeUL.parent = &newInteriorNode
+			if newInteriorNode.parent.leftChild == nodePointer {
+				newInteriorNode.parent.leftChild = &newInteriorNode
+			} else {
+				newInteriorNode.parent.rightChild = &newInteriorNode
+			}
+			nodePointer = &newInteriorNode
+			node.level--
+		case *interiorNode:
+			currentNodeI := nodePointer.(*interiorNode)
+			direction := getNthBit(key, position)
+
+			if direction { // go right
+				currentNodeI.rightHash = nil
+				if currentNodeI.rightChild == nil {
+					currentNodeI.rightChild = node
+					node.parent = currentNodeI
+					break insertLoop
+				} else {
+					nodePointer = currentNodeI.rightChild
+				}
+			} else { // go left
+				currentNodeI.leftHash = nil
+				if currentNodeI.leftChild == nil {
+					currentNodeI.leftChild = node
+					node.parent = currentNodeI
+					break insertLoop
+				} else {
+					nodePointer = currentNodeI.leftChild
+				}
+			}
+			position++
+		default:
+			return ErrInvalidTree
+		}
+	}
+	return nil
+}
+
+func (m *MerkleTree) RecomputeHash() {
+	m.computeHash([]bool{})
+}
+
+func (m *MerkleTree) computeHash(prefixBits []bool) {
+	if m.root.leftHash == nil {
+		m.root.leftHash = m.hashNode(
+			append(prefixBits, false), m.root.leftChild)
+	}
+	if m.root.rightHash == nil {
+		m.root.rightHash = m.hashNode(
+			append(prefixBits, true), m.root.rightChild)
+	}
+}
+
+func (m *MerkleTree) hashNode(prefixBits []bool, node interface{}) []byte {
+	if node == nil {
+		// empty node
+		return hashEmptyNode(m, prefixBits)
+	}
+
+	switch node.(type) {
+	case *interiorNode:
+		currentNodeI := node.(*interiorNode)
+		if currentNodeI.leftHash == nil {
+			currentNodeI.leftHash = m.hashNode(
+				append(prefixBits, false), currentNodeI.leftChild)
+		}
+		if currentNodeI.rightHash == nil {
+			currentNodeI.rightHash = m.hashNode(
+				append(prefixBits, true), currentNodeI.rightChild)
+		}
+		return hashInteriorNode(m, currentNodeI)
+	case *userLeafNode:
+		return hashLeafNode(m, node.(*userLeafNode))
+	}
+	return nil
+}
+
+func (m *MerkleTree) digest(input []byte) []byte {
+	h := m.hash.Hash
+	defer h.Reset()
+	h.Write(input)
+	return h.Sum(nil)
+}
+
+// Node manipulation functions
+
+var _ MerkleNode = (*userLeafNode)(nil)
+
+func (node *userLeafNode) Value() []byte {
+	return node.value
+}
+
+func (node *interiorNode) serialize() []byte {
+	var input []byte
+	input = append(input, node.leftHash...)
+	input = append(input, node.rightHash...)
+	return input
+}
+
+func leafNodeCommitment(m *MerkleTree, key string, value []byte) []byte {
+	commit := append([]byte{}, m.salt...)
+	commit = append(commit, key...)
+	commit = append(commit, value...)
+	return m.digest(commit)
```

Comment:
That's not so important as it only could be a optimization: But is there a reason for not using the hash functions `Write` (and Sum at the end) instead of copying all bytes into a new byte array?